internal/providers/aws/aws.go

```diff
@@ -73,3 +75,49 @@ func Init(f *resource.Fetcher) error {
 	f.S3RegionHint = regionHint
 	return nil
 }
+
+// FetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its configured, else
+// it will fall back to `IMDSv1`. By default, the value of url is an IMDSv1 endpoint
+// i.e. http://169.254.169.254/2019-10-01/user-data.
+func FetchFromAWSMetadata(u url.URL, opts resource.FetchOptions, f *resource.Fetcher) ([]byte, error) {
+	userdataLatestURL := url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "latest",
+	}
+	token, err := fetchAWSIMDSV2Token(userdataLatestURL, f)
+	if err == resource.ErrIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is unavailable; falling back to IMDSv1")
+	} else if err != nil {
+		return nil, err
+	} else if len(token) > 0 && err == nil {
+		opts.Headers = http.Header{
+			"X-aws-ec2-metadata-token": []string{token},
+		}
+		u.Path = path.Join(userdataLatestURL.Path, "user-data")
```

Comment:
We probably shouldn't override the passed-in path.  Do we need to use the version `latest`?  If the version in `userdataUrl` is too old, can we update it?

Reply:
I think we've recently updated the version using this PR: https://github.com/coreos/ignition/pull/993. It can be changed to the version `latest`, but we'd still need to override the passed-in path. Ideally, we'd want to keep the url common for both services i.e. `IMDSv1` and `IMDSv2`; something like:
```
	userdataUrl = url.URL{
		Scheme: "http",
		Host:   "169.254.169.254",
		Path:   "latest",
	}
```
However, for `IMDSv2`, we'd need to hit the following endpoint: https://github.com/coreos/ignition/pull/1154/files#diff-bf030bbb0152addedd720e7c13f083876784c0d85007fefb8e6bf75713b5376aR106 to get the session token, which forces us to use that option.

Reply:
Alright, as discussed off-line, the IMDSv2 token will be requested using the version `latest`, but we'd continue to fetch the user-data with the `2019-10-01` version.